datasets/flwr_datasets/partitioner/semantic_partitioner.py

```diff
@@ -0,0 +1,493 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Semantic partitioner class that works with Hugging Face Datasets."""
+
+
+import warnings
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+
+import datasets
+from flwr_datasets.common.typing import NDArrayFloat
+from flwr_datasets.partitioner.partitioner import Partitioner
+
+
+# pylint: disable=R0902, R0912, R0914
+class SemanticPartitioner(Partitioner):
```

Comment:
```suggestion
class ImageSemanticPartitioner(Partitioner):
```

Reply:
Since it works only with images, maybe it'll be more descriptive (I assume in the future there could be as well, e.g. `AudioSemanticPartitioner`. Or maybe some reference to the fact that labels are also considered can be included in the name (e.g., Supervised), but that might be too long, wdyt?

Reply:
Agree. Each type of data can have its own semantic information. So it seems practical to partition for other types of data as well. Perhaps the only difference is the way pretrained models are used to extract semantic information.

I will also change the python filenames to `image_semantic_partitioner.py` and `image_semantic_partitioner_test.py` as well.